File: app/src/main/java/com/qimian233/ztool/hook/modules/systemFramework/AiInputExpand.java
package com.qimian233.ztool.hook.modules.systemFramework;

import com.qimian233.ztool.hook.base.BaseHookModule;
import com.qimian233.ztool.hook.base.PreferenceHelper;

import de.robv.android.xposed.XC_MethodReplacement;
import de.robv.android.xposed.XposedHelpers;
import de.robv.android.xposed.callbacks.XC_LoadPackage;

/**
 * AI输入法扩展功能Hook模块
 * 功能：扩展AI触发符号，强制开启LGSI AI功能特性
 * 作用域：全局（动态检测类是否存在）
 */
public class AiInputExpand extends BaseHookModule {
    private final PreferenceHelper mPrefHelper = PreferenceHelper.getInstance();

    @Override
    public String getModuleName() {
        return "ai_input_expand"; // 模块唯一标识
    }

    @Override
    public String[] getTargetPackages() {
        return null; // 返回null，表示不限制特定包名，由supportsPackage控制
    }

    /**
     * 重写此方法以支持全局Hook
     * 因为RemoteInputConnectionImpl会在各个应用进程中加载
     */
    @Override
    public boolean supportsPackage(String packageName) {
        return true; // 允许所有包进入handleLoadPackage，我们在内部通过类是否存在来判断
    }

    @Override
    public void handleLoadPackage(XC_LoadPackage.LoadPackageParam lpparam) throws Throwable {
        // 核心逻辑分为两部分，分别放在try-catch块中，互不影响

        // 1. 修改 RemoteInputConnectionImpl 触发符号
        try {
            hookRemoteInputConnection(lpparam.classLoader);
        } catch (Throwable t) {
            // 某些进程可能没有这个类，属于正常现象，仅在调试时关注
            // logError("RemoteInputConnection hook failed in " + lpparam.packageName, t);
        }

        // 2. 强制开启 LgsiFeatures 功能
        try {
            hookLgsiFeatures(lpparam.classLoader);
        } catch (Throwable t) {
            // logError("LgsiFeatures hook failed in " + lpparam.packageName, t);
        }
    }

    private void hookRemoteInputConnection(ClassLoader classLoader) {
        String className = "android.view.inputmethod.RemoteInputConnectionImpl";

        // 检查类是否存在，不存在直接抛出/返回，避免无效Hook尝试
        Class<?> targetClass = XposedHelpers.findClassIfExists(className, classLoader);
        if (targetClass == null) return;

        // 定义新的触发符号数组，使用新的符号
        String[] newSignArray = mPrefHelper.getStringArray("AI_INPUT_EXPAND_SIGNS");

        // 修改静态常量数组 AI_COMMAND_SIGN_ARRAYS
        XposedHelpers.setStaticObjectField(targetClass, "AI_COMMAND_SIGN_ARRAYS", newSignArray);

        // 修改默认的 AI_COMMAND_SIGN
        XposedHelpers.setStaticObjectField(targetClass, "AI_COMMAND_SIGN", "&&");

        log("Successfully expanded AI input signs [&&] for package");
    }

    private void hookLgsiFeatures(ClassLoader classLoader) {
        String className = "com.lgsi.config.LgsiFeatures";

        Class<?> featureClass = XposedHelpers.findClassIfExists(className, classLoader);
        if (featureClass == null) return;

        // 强制 enabled 方法返回 true
        // 对应原代码：XposedHelpers.findAndHookMethod(featureClass, "enabled", int.class, ...)
        XposedHelpers.findAndHookMethod(
                featureClass,
                "enabled",
                int.class,
                new XC_MethodReplacement() {
                    @Override
                    protected Object replaceHookedMethod(MethodHookParam param) throws Throwable {
                        return true; // 强制返回 true
                    }
                }
        );
        log("Successfully forced LgsiFeatures check to TRUE");
    }
}

File: app/src/main/java/com/qimian233/ztool/hook/modules/systemframework/AiInputExpand.kt
package com.qimian233.ztool.hook.modules.systemframework

import android.annotation.SuppressLint
import android.text.TextUtils
import com.qimian233.ztool.data.keys.PreferenceKeys
import com.qimian233.ztool.hook.base.AppHookModule
import io.github.libxposed.api.XposedModuleInterface.PackageLoadedParam

/**
 * AI input method expansion hook module
 * Expands AI trigger signs and force-enables LGSI AI features
 * Scope: global (classes detected dynamically)
 */
@SuppressLint("PrivateApi")
class AiInputExpand : AppHookModule() {
    override fun getModuleName(): String = PreferenceKeys.AI_INPUT_EXPAND.name

    override fun getTargetPackages(): Array<String?>? = null

    /**
     * Overrides this method to support global hooking
     * because RemoteInputConnectionImpl loads in every app process
     */
    override fun supportsPackage(packageName: String?): Boolean = true

    override fun handleLoadPackage(param: PackageLoadedParam) {
        val classLoader = param.defaultClassLoader
        // 1. Modify RemoteInputConnectionImpl trigger signs
        runCatching { hookRemoteInputConnection(classLoader) }
        // 2. Force-enable LgsiFeatures features
        runCatching { hookLgsiFeatures(classLoader) }
    }

    private fun hookRemoteInputConnection(classLoader: ClassLoader) {
        val className = "android.view.inputmethod.RemoteInputConnectionImpl"

        // Check whether the class exists; return directly if not to avoid useless hook attempts
        val targetClass: Class<*>?
        try {
            targetClass = classLoader.loadClass(className)
        } catch (_: ClassNotFoundException) {
            return
        }

        // Define the new trigger sign array using the new signs
        val newSignArray = this.prefStringArray

        // Modify the static constant array AI_COMMAND_SIGN_ARRAYS
        findField(targetClass, "AI_COMMAND_SIGN_ARRAYS").set(null, newSignArray)

        // Modify the default AI_COMMAND_SIGN
        findField(targetClass, "AI_COMMAND_SIGN").set(null, "&&")

        logger.info("Successfully expanded AI input signs [&&] for package")
    }

    private fun hookLgsiFeatures(classLoader: ClassLoader) {
        val className = "com.lgsi.config.LgsiFeatures"

        val featureClass: Class<*>
        try {
            featureClass = classLoader.loadClass(className)
        } catch (_: ClassNotFoundException) {
            return
        }

        // Force the enabled method to return true
        try {
            val method = featureClass.getDeclaredMethod("enabled", Int::class.javaPrimitiveType)
            hookWithId(
                method,
                "lgsi_features_enabled"
            ) { true }
            logger.info("Successfully forced LGSI Features check to TRUE")
        } catch (_: NoSuchMethodException) {
            // Method not found; ignore
        }
    }

    private val prefStringArray: Array<String?>
        /**
         * Read comma-separated string array from preferences.
         */
        get() {
            val value: String? = try {
                remotePreferences.getString(PreferenceKeys.AI_INPUT_EXPAND_SIGNS.name, "")
            } catch (_: Throwable) {
                ""
            }
            if (TextUtils.isEmpty(value)) return arrayOfNulls(0)
            return value!!.split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        }
}
